### ISSUE TITLE
fix bomb

### ISSUE THREAD

**Issue by Mooses2k** [Mooses2k/SecretHistories#409]
* broken if blows up in hand; no damage if in world
* fix (get_parent.get_parent etc)

**Comment by Dokkae6949** [comment]
The problem which is causing the game to crash when the bomb explodes in the players hand seems to be in the 'Player_Animations.gd' script.

Specifically the first `if` statement in the `check_current_item_animation` function.
```swift
func check_current_item_animation():
	# This code checks the current item equipped by the player and updates the current_mainhand_item_animation to correspond to it 
	var main_hand_object = inventory.current_mainhand_slot
	var off_hand_object = inventory.current_offhand_slot

	if inventory.hotbar[main_hand_object] is GunItem or inventory.hotbar[off_hand_object] is GunItem : # <--- problem
```

Error message:
```
Left operand of 'is' was already freed.
```

My guess that either the object inside of the hotbar array was already freed (deleted)
or that the whole inventory already got deleted when the player died (thus the inventory.hotbar array does not exist anymore).

**Comment by Dokkae6949** [comment]
Kinda found the root cause (as expected) the bomb deletes itself without first removing itself from the inventory.

### Hacky Fix
Original code (`bomb.gd`):
```swift
func _process(delta):
	if countdown_started == true:
		$Explosion.lifetime -= delta
	if $Explosion.lifetime < 1.2:
		queue_free()
```

Hacky fix (`bomb.gd`):
```swift
func _process(delta):
	if countdown_started == true:
		$Explosion.lifetime -= delta
	if $Explosion.lifetime < 1.2:
		#queue_free()
		pass
```

Tho. __Don't use my fix!__ it just prevents the game from crashing at least.

### Suggested Fix
I would recommend adding some sort of helper functions like `safe_queue_free` on the inventory script which provides a simple way to remove itself first from the inventory and then `queue_free` the object after all the dependencies are cleared up.

###### (I'm not an expert so I still might be incorrect but at least this is what I believe is the cause)

**Comment by Dokkae6949** [comment]
Found a better hack to fix it.

Instead of removing `queue_free()` just out this snippet wherever it throws errors:
```swift
if not is_instance_valid(inventory.hotbar[main_hand_object]):
	return
```

This will then check if the reference (object) is even "valid" and if it isn't it does nothing.

Also I just found that the problem is not only present in one script but its seeded everywhere :/ also I found some minor code repetition in the `Player_Animation.gd` and `Character.gd` classes.

**Comment by Dokkae6949** [comment]
Okay I think I covered most locations which trigger this problem.

I ended up using this snipped at the top of each function in order to make sure that I don't block andy other logic which needs to be run:
```swift
if not is_instance_valid(item):
	item= null
```

**Comment by Dokkae6949** [comment]
Bomb explosion in players hand crashes game is now fixed (but hacky)

The next problem is that the bomb item does not get removed from the inventory after it explodes in the players hand (name still being displayed)
And then that the player is not getting damaged.

**Comment by Mooses2k** [comment]
closed with #418
